### DISCUSSION TITLE
flask SQLAlchemy db_migrate failing on large migration using SmallInteger

### DISCUSSION THREAD

**Discussion by markgalup** [miguelgrinberg/Flask-Migrate#480] [Q&A] [answered]
Hello Miguel,

I owe you a debt of thanks for this tremendous effort - I use your databases all day everyday. Thank you.

I've never run in to an issue with migrations before, but today I encountered many errors. I was added something like 60 columns, and out of those about 10-15 were the type **db.Column(db.SmallInteger, default=False)** Every time one of  those rows was processed, the migrate would throw an error. I checked my code for hours, but ultimately found that if I just kept running migrate, it would jump through these issues, and eventually migrated all columns. Tracebacks and pip freeze are below:

```
| => python ./db_migrate.py 
/Users/user/.pyenv/versions/2.7.18/envs/venv-py2/lib/python2.7/site-packages/sqlalchemy/dialects/mysql/mysqldb.py:176: Warning: 'utf8' is currently an alias for the character set UTF8MB3, but will be an alias for UTF8MB4 in a future release. Please consider using UTF8MB4 in order to be unambiguous.
  cursor.execute("SET NAMES %s" % charset_name)
Traceback (most recent call last):
  File "./db_migrate.py", line 21, in <module>
    api.upgrade(SQLALCHEMY_DATABASE_URI, SQLALCHEMY_MIGRATE_REPO)
  File "/Users/user/.pyenv/versions/2.7.18/envs/venv-py2/lib/python2.7/site-packages/migrate/versioning/api.py", line 186, in upgrade
    return _migrate(url, repository, version, upgrade=True, err=err, **opts)
  File "<string>", line 2, in _migrate
  File "/Users/user/.pyenv/versions/2.7.18/envs/venv-py2/lib/python2.7/site-packages/migrate/versioning/util/__init__.py", line 167, in with_engine
    return f(*a, **kw)
  File "/Users/user/.pyenv/versions/2.7.18/envs/venv-py2/lib/python2.7/site-packages/migrate/versioning/api.py", line 366, in _migrate
    schema.runchange(ver, change, changeset.step)
  File "/Users/user/.pyenv/versions/2.7.18/envs/venv-py2/lib/python2.7/site-packages/migrate/versioning/schema.py", line 93, in runchange
    change.run(self.engine, step)
  File "/Users/user/.pyenv/versions/2.7.18/envs/venv-py2/lib/python2.7/site-packages/migrate/versioning/script/py.py", line 154, in run
    script_func(engine)
  File "/Users/user/dev/ad_database_CURRENT/db_repository/versions/030_migration.py", line 207, in upgrade
    post_meta.tables['audiobook'].columns['qc_eng_3_complete'].create()
  File "/Users/user/.pyenv/versions/2.7.18/envs/venv-py2/lib/python2.7/site-packages/migrate/changeset/schema.py", line 595, in create
    stmt = table.update().values({self: engine._execute_default(self.default)})
AttributeError: 'NoneType' object has no attribute 'update'
```

```
appdirs==1.4.3
Babel==1.3
bcrypt==3.1.7
blinker==1.3
cffi==1.15.0
click==7.1.2
configparser==4.0.2
contextlib2==0.6.0.post1
coverage==3.7.1
decorator==3.4.2
Flask==1.1.4
Flask-Babel==0.9
Flask-Bcrypt==0.6.2
Flask-Login==0.2.11
Flask-Mail==0.9.1
Flask-OpenID==1.2.4
Flask-SQLAlchemy==2.5.1
Flask-Uploads==0.2.1
Flask-WhooshAlchemy==0.56
Flask-WTF==0.14.3
flipflop==1.0
guess-language==0.2
importlib-metadata==2.1.3
itsdangerous==0.24
Jinja2==2.11.3
MarkupSafe==0.23
MySQL-python==1.2.5
packaging==16.8
pathlib2==2.3.7.post1
pbr==5.9.0
pycparser==2.21
pyparsing==2.2.0
python-bcrypt==0.3.1
python-openid==2.2.5
pytz==2015.4
scandir==1.10.0
schedule==0.5.0
six==1.9.0
speaklater==1.3
SQLAlchemy==1.4.36
sqlalchemy-migrate==0.13.0
sqlparse==0.1.15
Tempita==0.5.2
typing==3.10.0.0
Unidecode==1.0.23
Werkzeug==1.0.1
Whoosh==2.7.0
WTForms==2.0.2
zipp==1.2.0
```

I'm reluctant to push this to my production server in case there are more significant issues. I appreciate any help very much.

Best,
MG

**Comment by miguelgrinberg** [comment, marked as answer]
This is probably an Alembic issue. Flask-Migrate doesn't do any of this, I just take care of the integration of Alembic with Flask. You may want to create an example that reproduces the problem and get advice from the Alembic team on it.

**Reply by markgalup** [reply]
Great to know and understood. Thank you very much, Miguel. A few quick questions:

- Do you have any specific thoughts on where exactly I should let Alembic know about this? Is there a specific site/direction you take to discuss their integration with your software?
- Do you think there are any concerns if I just run flask migrate until it goes fully through, sort of brute forcing through this bug? I checked the MySQL DB and it all seems to be creating correctly.

Thank you again,
MG

**Reply by miguelgrinberg** [reply]
Alembic discussion board: https://github.com/sqlalchemy/alembic/discussions

I would not keep a broken migration on my project, especial when the reason why the migration fails is not understood.

**Reply by markgalup** [reply]
Thank you, sir!

**Reply by markgalup** [reply]
I just asked this on the Alembic discussion board. One last quick question, as I'm relatively new to Github Discussions/Issues - would the smartest move have been to use the "Transfer this Discussion" functionality in the sidebar here, or is that considered sort of in poor taste to drag things into new threads?

**Reply by miguelgrinberg** [reply]
Oh, I wasn't even aware that GitHub allows you to transfer the discussion to another project. If I was on the receiving project's side I would prefer a new discussion with a concise description of the issue, and a reference to the other discussion for background. But that's just me, other people me prefer a transfer.

**Reply by miguelgrinberg** [reply]
As a side note, you have ignored my suggestion that you create a small example that reproduces the problem and present that. This exercise would have helped you make sure the problem is actually in Alembic and not in your own application. It would also make it easy for the maintainers to evaluate the bug without having to write an application from scratch and hope it reproduces the problem.

After looking at Mike's reply, I think writing a test application would have helped you realize that you are not even using Flask-Migrate nor Alembic.

**Reply by markgalup** [reply]
Thank you for both of these messages, Miguel! Confirming that I didn't ignore your suggestion, I've just been trying to figure out how to do it. I'm not an expert coder and am working to keep my business running while I'm troubleshooting this situation, so it just takes me a bit longer.

Regarding your last message, I had been wondering about this recently. When I migrate, I'm executing the python script manually (going to the terminal and typing python ./db_migrate.py) within pycharm, with the venv activated. Should I be using a different method? This is the only way I've ever known to do it.

Thank you VERY much for your attention and assistance.

Best,
Mark

**Reply by miguelgrinberg** [reply]
You are using and ancient and pretty much unmaintained package to do your migrations. I missed this in the stack trace, but you posted your question here and I assumed you were using Flask-Migrate. You are using something else, so I cannot really comment since I don't know what you are using beyond the fact that it is very old tech.

**Reply by markgalup** [reply]
Ok great, understood. I will go in to more detail later, but would you be willing to do a payed zoom consultation with me? 

I rely on your database tutorial build, and the reason it's outdated is due to me following your tutorial to build this 7 years ago and needing to update. I have specific questions regarding that. 

Please let me know how I could email/connect with you, or if you'd prefer my email address. I would greatly appreciate it and I will not waste your time. 

Best,
Mark Galup

**Reply by miguelgrinberg** [reply]
Oh, I see. You took code from the 2012 version of the tutorial, the original. It's 10 years old now. I have created Flask-Migrate to replace that migration method a long time ago. Unfortunately I can't really help with this. My recommendation is that you switch to Flask-Migrate, as you are unlikely to find any support for sqlalchemy-migrate these days, it's been long forgotten.

The Flask-Migrate documentation should help you get up and running, but in case you are wondering, there is no mechanism to transfer your existing migrations to Flask-Migrate/Alembic, you will have to start a new migration repository from scratch.

**Reply by markgalup** [reply]
Thank you for this, Miguel. I'm glad at least this is making sense now, and apologies for not understanding the context properly here. 

First, confirming I will switch to Flask-Migrate. To clarify, are you saying that I will just lose my ability to traverse through versions of migrations? Or are you saying my underlying database will be affected?

Second, just want to say that I run a business with help from this database, so if you'd ever generally be willing to do a consultation, I would love to know if that's a possibility. Currently, I have no support system to discuss the codebase with.  

Thanks again,
Mark

**Reply by miguelgrinberg** [reply]
What I'm saying is that Flask-Migrate and Alembic are not going to see the database migrations that you have created so far. You will only be able to track database migrations going forward with these tools. I think the most reasonable way to go forward in your case would be to toss away your migration history, and start fresh with Flask-Migrate, if that is feasible. I think if you use the documentation plus my blog post "How to add Flask-Migrate to an existing project"  you should be able to do it.

If you need your current migrations to remain available, then you are going to need some sort of hybrid solution that uses the old and the new migration tooling. I do not recommend this.

I'm currently not offering consultations, I don't have time unfortunately. If/when I decide to restart, I'll update my blog with the link to make appointments.

**Reply by markgalup** [reply]
Sounds great - thank you very much to all.

I'm almost fully implemented with Flask-Migrate now. I have one question as I'd love to have a super clean and healthy start to the migrations:

I am using MySQL specific conditions, so I created an empty database I called "delete" and then tried to set the DATABASE_URL variable to it. However, upon running flask db migrate, it seems consistently to be looking at my current database. 

I say this because I believe that Flask-Migrate is noticing some incorrect documentation in my Model, and correcting it, but I'm ONLY seeing that, not any of the other elements of the Model being created. Here's the code for context:

```

| => DATABASE_URL=jdbc:mysql://localhost/delete flask db migrate
/Users/markgalup/.pyenv/versions/2.7.18/envs/venv-py2/lib/python2.7/site-packages/sqlalchemy/dialects/mysql/mysqldb.py:176: Warning: 'utf8' is currently an alias for the character set UTF8MB3, but will be an alias for UTF8MB4 in a future release. Please consider using UTF8MB4 in order to be unambiguous.
  cursor.execute("SET NAMES %s" % charset_name)
INFO  [alembic.runtime.migration] Context impl MySQLImpl.
INFO  [alembic.runtime.migration] Will assume non-transactional DDL.
INFO  [alembic.autogenerate.compare] Detected removed table u'migrate_version'
INFO  [alembic.autogenerate.compare] Detected removed foreign key (cue_sheet_engineer_id)(id) on table audiobook
INFO  [alembic.autogenerate.compare] Detected removed foreign key (edit_engineer_id)(id) on table audiobook
INFO  [alembic.autogenerate.compare] Detected removed foreign key (assembly_engineer_id)(id) on table audiobook
INFO  [alembic.autogenerate.compare] Detected removed foreign key (sound_designer_id)(id) on table audiobook
INFO  [alembic.autogenerate.compare] Detected removed foreign key (archive_engineer_id)(id) on table audiobook
INFO  [alembic.autogenerate.compare] Detected removed foreign key (qc_engineer_id)(id) on table audiobook
INFO  [alembic.autogenerate.compare] Detected removed foreign key (multicast_engineer_id)(id) on table audiobook
INFO  [alembic.autogenerate.compare] Detected removed foreign key (mastering_engineer_id)(id) on table audiobook
INFO  [alembic.autogenerate.compare] Detected added foreign key (cue_sheet_engineer_id)(id) on table audiobook
INFO  [alembic.autogenerate.compare] Detected added foreign key (archive_engineer_id)(id) on table audiobook
INFO  [alembic.autogenerate.compare] Detected added foreign key (assembly_engineer_id)(id) on table audiobook
INFO  [alembic.autogenerate.compare] Detected added foreign key (comp_engineer_id_3)(id) on table audiobook
INFO  [alembic.autogenerate.compare] Detected added foreign key (comp_engineer_id_2)(id) on table audiobook
INFO  [alembic.autogenerate.compare] Detected added foreign key (sound_designer_id)(id) on table audiobook
INFO  [alembic.autogenerate.compare] Detected added foreign key (qc_engineer_id_3)(id) on table audiobook
INFO  [alembic.autogenerate.compare] Detected added foreign key (mastering_engineer_id)(id) on table audiobook
INFO  [alembic.autogenerate.compare] Detected added foreign key (edit_engineer_id_3)(id) on table audiobook
INFO  [alembic.autogenerate.compare] Detected added foreign key (qc_engineer_id_2)(id) on table audiobook
INFO  [alembic.autogenerate.compare] Detected added foreign key (comp_engineer_id)(id) on table audiobook
INFO  [alembic.autogenerate.compare] Detected added foreign key (multicast_engineer_id)(id) on table audiobook
INFO  [alembic.autogenerate.compare] Detected added foreign key (edit_engineer_id)(id) on table audiobook
INFO  [alembic.autogenerate.compare] Detected added foreign key (edit_engineer_id_2)(id) on table audiobook
INFO  [alembic.autogenerate.compare] Detected added foreign key (qc_engineer_id)(id) on table audiobook
INFO  [alembic.autogenerate.compare] Detected added foreign key (sound_designer_id_2)(id) on table audiobook
  Generating /Users/markgalup/mg-dev/database_CURRENT/migrations/versions/4924c4644e4e_.py ...  done

```

Does this look like a correct usage of a MySQL URL? `jdbc:mysql://localhost/delete flask db migrate`

EDIT: I've confirmed the above is correct. I just tried the DATABASE_URL command with sqlite:/// and it gives the same output as above. I think the DATABASE_URL variable is not getting parsed possibly. Thoughts?

Thank you again,
Mark

**Reply by miguelgrinberg** [reply]
The fact that the migration detects things to delete indicates that this `delete` database that you are using is not empty. Is it supposed to be empty? I assume you are following my instructions on generating a first migration from an empty database here? You should make sure your database is really empty when you do this. You should only see added stuff and it should be everything getting added, all your tables, indexes, etc. Nothing should be deleted since there shouldn't be anything at all in the temporary database.

**Reply by markgalup** [reply]
Thanks, Miguel. I can unequivocally guarantee that the delete database was empty. I created it just before that statement. 

I also tried several other things and it returned the same statement, for example, sqlite:///. For whatever reason, that variable isn't passing into the flask db migrate command. 

Thoughts?

Thank you!
Mark

**Reply by miguelgrinberg** [reply]
Flask-Migrate uses your Flask and Flask-SQLAlchemy configuration to get the database, so this is likely a configuration issue that you need to debug in your application. My guess is that the changes that you are making to DATABASE_URL are also not being applied on your Flask application, which would confirm this is a general configuration problem.

**Reply by markgalup** [reply]
Ok got it - that makes sense. I'll do more digging, but it's sounding like maybe you're saying that passing the DATABASE_URL in the CLI command here is overriding a variable that's built in to the Flask app elsewhere. I believe I remember seeing DATABASE_URI, so I'll confirm. 

Thanks!
Mark

**Reply by miguelgrinberg** [reply]
`DATABASE_URL` is an application specific environment variable that I normally use in my applications. If you haven't done anything in the application to read this variable and install it in the Flask-SQLAlchemy configuration, then setting this variable is not going to do anything.

**Reply by markgalup** [reply]
Ah ok, I think I am realizing my misconception. I was thinking the `DATABASE_URL` variable was self-contained in the Flask-Migrate package. I'll make sure the application is creating the environment variable and then see how this goes. 

Thanks!
Mark

**Reply by markgalup** [reply]
Miguel, confirming that I resolved this by establishing the environment var, and it's all working correctly now! Thank you very much again for all your help. 

Best,
Mark

**Reply by miguelgrinberg** [reply]
Awesome, congrats!
